src/data/set/function.lean

```diff
@@ -59,10 +59,13 @@ has codomain `↥s` instead of `subtype s`. -/
 def cod_restrict (f : α → β) (s : set β) (h : ∀ x, f x ∈ s) : α → s :=
 λ x, ⟨f x, h x⟩
 
-@[simp] lemma coe_cod_restrict_apply (f : α → β) (s : set β) (h : ∀ x, f x ∈ s) (x : α) :
+@[simp] lemma coe_cod_restrict_apply {f : α → β} {s : set β} (h : ∀ x, f x ∈ s) (x : α) :
   (cod_restrict f s h x : β) = f x :=
 rfl
 
+lemma factor_comp_cod_restrict {f : α → β} {g : β → γ} {b : set β}
```

Comment:
strange lemma name (why `factor` instead of `restrict`? Also, I'd probably switch the two sides in the conclusion of the lemma, to make the right hand side the simpler one.

Reply:
Because this statement would be phrased this way in CT: if a morphism factors in two morphisms of which the first has image, then it has a factorization in two morphisms of which the first is the standard morphism associated to the morphism with image (I'd guess at least). I can change it to restrict. I'd also leave this order because this is the true meaning of the lemma and I don't think it will ever be used the other way around.

Reply:
Predictable lemma names are better because they are much easier to find (someone knowing about `restrict` and `cod_restrict` will never look for a name containing `factor`). So yes, please change (and also please change the order, just for the sake of coherence with the rest of the library where the simpler form is always on the right).

Reply:
Ok I changed the name as you suggested, as for the order I remain convinced it will never be used in the other sense and it is not a simp lemma: it is often the case but it is not true that the simpler form it is always on the right for non-simp lemmas. I can quote you many examples in mathlib where this is not the case, also from your own code.

Reply:
I don't want to be rude here, but I am not sure that you realize that reviewing time is time given to the community, and that this kind of answer is not very encouraging in this respect.

Reply:
I am really sorry if my answer sounded disrespectful, I did not mean it to be. I do realize that and I am very grateful for all the time you took for reviewing my code and I am very aware of how fundamental your help was each time. I know this detail sounds stupid, and it's not worth spending so much time on it, but since I was not convinced by what you said, I just wanted to say why. But just so that you know, I am aware this is a non-important detail, and the fact that I was not convinced here does not dent the esteem I have for your ideas and your suggestions and the respect I have for you.

If I did not expand enough on why I am not convinced, the reason is that I don't think this will not be used as a simp lemma, since I don't see when someone could happen in the situation where this would be needed to simplify. If someone happens to be in that situation it will probably be because they want it to happen. So, to avoid using this lemma always with a left arrow, I don't think changing the order is a good idea. I probably phrased it badly before, but I still believe this has happened in mathlib already.

Again, I do realize this is not very important, so if you are not convinced and do not want to discuss this, you can say so and I will accept, but if you want to convince me with argumentations, I will answer with argumentations until I am convinced, even if it is about the stupid details (this is true in general: if in the future you will want me to do something about a non-important detail and you will not want to lose time to discuss, you can just say so, and I will only refuse if I think the detail is actually important).

Reply:
There are two competing conventions here:
(a) lemmas should be written so that, when rewriting, one uses most often the direct direction.
(b) lemmas should be written with the most complicated term on the left.
When these two conventions are contradictory, the meta-convention is that (b) wins in mathlib. We have indeed several lemmas that are used 100% of the time in rewrites from right to left.